### PR TITLE
refactor: consolidate `MessageData` constructors into lazy formatting with infos

### PR DESCRIPTION
**Pull request by Kha**
In preparation of #375

### CODE REVIEW

File: src/Lean/Message.lean [hunk 1]
```diff
@@ -37,16 +38,22 @@ structure NamingContext where
   currNamespace : Name
   openDecls : List OpenDecl
 
+/-- Lazily formatted text to be used in `MessageData`. -/
+structure PPFormat where
+  /-- Pretty-prints text using surrounding context, if any. -/
+  pp : Option PPContext → IO FormatWithInfos
+  /-- Searches for synthetic sorries in original input. Used to filter out certain messages. -/
+  hasSyntheticSorry : MetavarContext → Bool := fun _ => false
```
Comment (Kha) [inline src/Lean/Message.lean:46]: We could probably get rid of this field in favor of "check if `pp` returns a `TermInfo` at the root and use its `expr`" in exchange for formatting terms we will not display

**Review by gebner — APPROVED** [review]
This is the exactly the design we discussed in #375, so I've got nothing to complain about.  Should we do a stage0 update?

File: src/Lean/Widget/InteractiveDiagnostic.lean [hunk 2]
```diff
@@ -22,7 +22,7 @@ structure LazyTraceChildren where
   deriving TypeName
 
 inductive MsgEmbed where
-  | expr : CodeWithInfos → MsgEmbed
+  | code : CodeWithInfos → MsgEmbed
```
[the remark this answers is not in the document]
Comment (gebner) [reply src/Lean/Widget/InteractiveDiagnostic.lean:25]:
```suggestion
  | expr : CodeWithInfos → MsgEmbed
```

Yeah, Wojciech is completely right.  This should really stay `expr`.

File: src/Lean/Widget/InteractiveDiagnostic.lean [hunk 3]
```diff
@@ -96,15 +96,10 @@ where
 
   go : NamingContext → Option MessageDataContext → MessageData → MsgFmtM Format
   | _,    _,         ofFormat fmt             => withIgnoreTags (pure fmt)
-  | _,    _,         ofLevel u                => return format u
-  | _,    _,         ofName n                 => return format n
-  | nCtx, some ctx,  ofSyntax s               => withIgnoreTags (ppTerm (mkPPContext nCtx ctx) ⟨s⟩) -- HACK: might not be a term
```
Comment (Vtec234) [inline src/Lean/Widget/InteractiveDiagnostic.lean:101]: This change looks a bit fishy since `withIgnoreTags` is no longer used.

Comment (Vtec234) [reply src/Lean/Widget/InteractiveDiagnostic.lean:101]: Okay, I convinced myself that this is fine, but man this code is awfully complicated lol. I wrote a docstring trying to explain what's going on and pushed it directly here. Hope that's okay, but do let me know if you'd rather I not push directly.

File: src/Lean/Widget/InteractiveDiagnostic.lean [hunk 4]
```diff
@@ -55,31 +60,45 @@ private def mkPPContext (nCtx : NamingContext) (ctx : MessageDataContext) : PPCo
...
+`Format` corresponds to an `EmbedFmt`. The kind of this tag determines how the nested `Format`
+object (possibly including further `.tag`s), is processed. For example, if the output is
+`.tag (.expr ctx infos) fmt` then tags in the nested `fmt` object refer to elements of `infos`.
+
+In the second stage, we recursively transform such a `Format` into `TaggedText MsgEmbed` according
+to the rule above by first pretty-printing it and then grabbing data referenced by the tags from
+all the nested arrays (such as the `infos` array in the example above).
+
+We cannot easily do the translation in a single `MessageData → TaggedText MsgEmbed` step because
+that would effectively require reimplementing the (stateful, to keep track of indentation)
```
Comment (Kha) [inline src/Lean/Widget/InteractiveDiagnostic.lean:76]: @Vtec234 The pretty printer does not track indentation. Only `Format.pretty` does, but I guess that's not what you mean here.

File: src/Lean/Widget/InteractiveDiagnostic.lean [hunk 5]
```diff
@@ -94,55 +113,46 @@ where
...
         let children := children.map fun child =>
           MessageData.withNamingContext nCtx <|
             match ctx with
             | some ctx => MessageData.withContext ctx child
             | none     => child
-        pure (.lazy children)
+        .lazy children
       else
-        pure (.strict (← children.mapM (go nCtx ctx)))
+        .strict (← children.mapM (go nCtx ctx))
```
Comment (gebner) [inline src/Lean/Widget/InteractiveDiagnostic.lean:143]: This change had a subtle performance impact: #1987

**Pull request merged**